packages/contracts-bedrock/scripts/L2Genesis.s.sol

```diff
@@ -318,112 +389,103 @@ contract L2Genesis is Script, Artifacts {
+
+        address eas;
+        assembly {
+            eas := create(0, add(code, 0x20), mload(code))
+        }
 
-    /// @dev Returns true if the address is a predeploy.
-    function _isDefinedPredeploy(address _addr) internal pure returns (bool) {
-        return _addr == Predeploys.L2_TO_L1_MESSAGE_PASSER || _addr == Predeploys.L2_CROSS_DOMAIN_MESSENGER
-            || _addr == Predeploys.L2_STANDARD_BRIDGE || _addr == Predeploys.L2_ERC721_BRIDGE
-            || _addr == Predeploys.SEQUENCER_FEE_WALLET || _addr == Predeploys.OPTIMISM_MINTABLE_ERC20_FACTORY
-            || _addr == Predeploys.OPTIMISM_MINTABLE_ERC721_FACTORY || _addr == Predeploys.L1_BLOCK_ATTRIBUTES
-            || _addr == Predeploys.GAS_PRICE_ORACLE || _addr == Predeploys.DEPLOYER_WHITELIST || _addr == Predeploys.WETH9
-            || _addr == Predeploys.L1_BLOCK_NUMBER || _addr == Predeploys.LEGACY_MESSAGE_PASSER
-            || _addr == Predeploys.PROXY_ADMIN || _addr == Predeploys.BASE_FEE_VAULT || _addr == Predeploys.L1_FEE_VAULT
-            || _addr == Predeploys.GOVERNANCE_TOKEN || _addr == Predeploys.SCHEMA_REGISTRY || _addr == Predeploys.EAS;
-    }
+        console.log("Setting %s implementation at: %s", cname, impl);
+        vm.etch(impl, eas.code);
 
-    /// @dev Function to compute the expected address of the predeploy implementation
-    ///      in the genesis state.
-    function _predeployToCodeNamespace(address _addr) internal pure returns (address) {
-        return address(
-            uint160(uint256(uint160(_addr)) & 0xffff | uint256(uint160(0xc0D3C0d3C0d3C0D3c0d3C0d3c0D3C0d3c0d30000)))
-        );
+        /// Reset so its not included state dump
+        vm.etch(address(eas), "");
+        vm.resetNonce(address(eas));
     }
 
-    function _setImplementationCode(address _addr, string memory _name) internal returns (address) {
-        address impl = _predeployToCodeNamespace(_addr);
-        console.log("Setting %s implementation at: %s", _name, impl);
-        vm.etch(impl, vm.getDeployedCode(string.concat(_name, ".sol:", _name)));
+    /// @dev Sets all the preinstalls.
+    function setPreinstalls() internal {
+        _setPreinstallCode(Preinstalls.MultiCall3);
+        _setPreinstallCode(Preinstalls.Create2Deployer);
+        _setPreinstallCode(Preinstalls.Safe_v130);
+        _setPreinstallCode(Preinstalls.SafeL2_v130);
+        _setPreinstallCode(Preinstalls.MultiSendCallOnly_v130);
+        _setPreinstallCode(Preinstalls.SafeSingletonFactory);
+        _setPreinstallCode(Preinstalls.DeterministicDeploymentProxy);
+        _setPreinstallCode(Preinstalls.MultiSend_v130);
+        _setPreinstallCode(Preinstalls.Permit2);
+        _setPreinstallCode(Preinstalls.SenderCreator);
+        _setPreinstallCode(Preinstalls.EntryPoint); // ERC 4337
+        _setPreinstallCode(Preinstalls.BeaconBlockRoots);
+    }
 
-        _verifyProxyImplementationAddress(_addr, impl);
+    function activateEcotone() public {
```

Comment:
This technically doesn't fully activate ecotone, we would be able to make good use of https://github.com/foundry-rs/foundry/pull/4931 to accomplish this easily. There are a few paths forward here:

- get the cheatcode into forge and just hardcode the rlp encoded tx. This would also require building foundry with optimism feature flag since its a deposit tx
- defining a solidity version of the "message" that is represented by the upgrade deposit tx

I lean towards just defining the messages given we would need to use a special foundry with optimism build flag and its unclear if that would even work e2e. Not ideal that its not 1:1 with the go, we would prob want a differential test long term

Reply:
The upgrade txs don't apply, as it deploys the same contracts that are already in place by default. The 4788 contract is installed as preinstall, independent of hardfork. And then only the `setL1BlockValuesEcotone` call remains to do here.